Update the LLM Agents guide with Gemini2.0 Thinking demo (#10247)
gradio-app/gradio: guides/05_chatbots/03_agents-and-tool-usage.md

Tags: LLM, AGENTS, CHAT
Related spaces: https://huggingface.co/spaces/gradio/agent_chatbot, https://huggingface.co/spaces/gradio/langchain-agent

> The Gradio Chatbot can natively display intermediate thoughts and tool usage. This makes it perfect for creating UIs for LLM agents and Thinking LLMs. This guide will show you how.

Reviewer: ```suggestion
The Gradio Chatbot can natively display intermediate thoughts and tool usage. This makes it perfect for creating UIs for LLM agents and chain-of-thought (CoT) demos. This guide will show you how to display intermediate thoughts.
```

That's it! See our finished langchain demo [here](https://huggingface.co/spaces/gradio/langchain-agent).

> ## Building with Visible Thinking LLMs

Reviewer: ```suggestion
## Chain-of-Thought (CoT) demos
```

Author: I believe the terms 'Thinking' and 'Reasoning' will be increasingly popular for LLMS. Therefore, I want to ensure that our messaging is aligned. In my opinion, 'thinking' is currently more commonly used than 'Chain-of-thought.' In fact, Google has referred to its model as 'Gemini2.0 Flash Thinking Experimental,' so adopting 'Thinking' seems more logical to me.

The Gradio Chatbot can natively display intermediate thoughts of a _thinking_ LLM. This makes it perfect for creating UIs that show how an AI model "thinks" while generating responses. Below guide will show you how to build a chatbot that displays Gemini AI's thought process in real-time.

> ### A real example using Gemini2.0 Flash Thinking API

Reviewer: ```suggestion
### A real example using Gemini 2.0 Flash Thinking API
```

> Let's create a complete chatbot that shows its thoughts and responses in real-time. We'll use Google's Gemini API for accessing Gemini2.0 Flash Thinking LLM and Gradio for the UI.

Reviewer: ```suggestion
Let's create a complete chatbot that shows its thoughts and responses in real-time. We'll use Google's Gemini API for accessing Gemini 2.0 Flash Thinking LLM and Gradio for the UI.
```

> We'll begin with imports and setting up the gemini client. Note that you'll need to acquire a Google Gemini API key first -

Reviewer: Perhaps you can link to where one can obtain a Gemini API key
```suggestion
We'll begin with imports and setting up the gemini client. Note that you'll need to acquire a Google Gemini API key first -
```

Author: Good idea! Will add a hyperlink.

            messages[-1] = ChatMessage(
                role="assistant",
                content=thought_buffer,
                metadata={"title": "⏳Thinking: *The thoughts produced by the Gemini2.0 Flash model are experimental"}
            )
            
            # Add response message
            messages.append(
                ChatMessage(
                    role="assistant",
                    content=parts[1].text
                )
            )
            thinking_complete = True
            
        elif thinking_complete:
            # Continue streaming response
            response_buffer += current_chunk
            messages[-1] = ChatMessage(
                role="assistant",
                content=response_buffer
            )
            
        else:
            # Continue streaming thoughts
            thought_buffer += current_chunk
            messages[-1] = ChatMessage(
                role="assistant",
                content=thought_buffer,
                metadata={"title": "⏳Thinking: *The thoughts produced by the Gemini2.0 Flash model are experimental"}
            )
        
        yield messages
```

> Then, let's create the Gradio interface:

Reviewer: Could we use `gr.ChatInterface` instead? `gr.ChatInterface` now supports the function returning multiple messages (you just return or yield a `list`), which would simplify the Gradio code and provide a better UI.

Author: I was thinking that a more granular implementation with `gr.Chatbot` would add more value to the guide. But there are no strong opinions here.

```python
with gr.Blocks() as demo:
    gr.Markdown("# Chat with Gemini 2.0 Flash and See its Thoughts 💭")
    
    chatbot = gr.Chatbot(
        type="messages",
        label="Gemini2.0 'Thinking' Chatbot",
        render_markdown=True,
    )
    
    input_box = gr.Textbox(
        lines=1,
        label="Chat Message",
        placeholder="Type your message here and press Enter..."
    )
    
    # Set up event handlers
    msg_store = gr.State("")  # Store for preserving user message
    
    input_box.submit(
        lambda msg: (msg, msg, ""),  # Store message and clear input
        inputs=[input_box],
        outputs=[msg_store, input_box, input_box],
        queue=False
    ).then(
        user_message,  # Add user message to chat
        inputs=[msg_store, chatbot],
        outputs=[input_box, chatbot],
        queue=False
    ).then(
        stream_gemini_response,  # Generate and stream response
        inputs=[msg_store, chatbot],
        outputs=chatbot
    )

demo.launch()
```

This creates a chatbot that:

- Displays the model's thoughts in a collapsible section
- Streams the thoughts and final response in real-time
- Maintains a clean chat history

>  That's it! You now have a chatbot that not only responds to users but also shows its thinking process, creating a more transparent and engaging interaction. See our finished Gemini2.0 Flash Thinking demo [here](https://huggingface.co/spaces/ysharma/Gemini2-Flash-Thinking).

Reviewer: ```suggestion
 That's it! You now have a chatbot that not only responds to users but also shows its thinking process, creating a more transparent and engaging interaction. See our finished Gemini 2.0 Flash Thinking demo [here](https://huggingface.co/spaces/ysharma/Gemini2-Flash-Thinking).
```
